Tests/Tests/STPPaymentMethodsViewControllerTest.m

```diff
@@ -0,0 +1,172 @@
+@end
+
+@implementation STPPaymentMethodsViewControllerTest
+
+- (STPPaymentMethodsViewController *)buildViewControllerWithCustomer:(STPCustomer *)customer
+                                                       configuration:(STPPaymentConfiguration *)config
+                                                            delegate:(id<STPPaymentMethodsViewControllerDelegate>)delegate {
+    STPTheme *theme = [STPTheme defaultTheme];
+    STPCustomerContext *mockCustomerContext = [STPMocks staticCustomerContextWithCustomer:customer];
+    STPPaymentMethodsViewController *vc = [[STPPaymentMethodsViewController alloc] initWithConfiguration:config
+                                                                                                   theme:theme
+                                                                                         customerContext:mockCustomerContext
+                                                                                                delegate:delegate];
+    if (vc) {
+        XCTAssertNotNil(vc.view);
+    }
+    return vc;
+}
+
+/**
+ When the customer has no sources, and card is the sole available payment 
+ method, STPAddCardViewController should be shown.
+ */
+- (void)testInitWithNoSourcesAndConfigWithUseSourcesOffAndCardAvailable {
+    STPCustomer *customer = [STPFixtures customerWithNoSources];
+    STPPaymentConfiguration *config = [STPFixtures paymentConfiguration];
+    config.additionalPaymentMethods = STPPaymentMethodTypeNone;
+    id<STPPaymentMethodsViewControllerDelegate>delegate = OCMProtocolMock(@protocol(STPPaymentMethodsViewControllerDelegate));
+    STPPaymentMethodsViewController *sut = [self buildViewControllerWithCustomer:customer
+                                                                   configuration:config
+                                                                        delegate:delegate];
+    XCTAssertTrue([sut.internalViewController isKindOfClass:[STPAddCardViewController class]]);
+}
+
+/**
+ When the customer has a single card token source and the available payment methods
+ are card and apple pay, STPPaymentMethodsInternalVC should be shown.
+ */
+- (void)testInitWithSingleCardTokenSourceAndCardAvailable {
+    STPCustomer *customer = [STPFixtures customerWithSingleCardTokenSource];
+    STPPaymentConfiguration *config = [STPFixtures paymentConfiguration];
+    config.additionalPaymentMethods = STPPaymentMethodTypeAll;
+    id<STPPaymentMethodsViewControllerDelegate>delegate = OCMProtocolMock(@protocol(STPPaymentMethodsViewControllerDelegate));
+    STPPaymentMethodsViewController *sut = [self buildViewControllerWithCustomer:customer
+                                                                   configuration:config
+                                                                        delegate:delegate];
+    XCTAssertTrue([sut.internalViewController isKindOfClass:[STPPaymentMethodsInternalViewController class]]);
```

Comment:
can we also check that the internal `paymentMethods` and `selectedPaymentMethod` props are set to the expected values? 

alternatively, we could add some tests for `filteredSourceTupleForUIWithConfiguration`, and avoid exposing those internal properties to tests.

Reply:
I'll just add some tests for filter 👍